modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+        for (unsigned int i = 0; i < cut_elem->n_nodes(); ++i)
+        {
+          tri.push_back(cut_elem->node(i));
+        }
+        myfile << tri[0] << " " << tri[1] << " " << tri[2] << std::endl;
+      }
+    }
+    myfile.close();
+  }
+}
+
+MeshCut3DUserObject::~MeshCut3DUserObject() {}
+
+bool
+MeshCut3DUserObject::active(Real time) const
+{
+  // TBD
+  return time >= 0;
+}
+
+bool
+MeshCut3DUserObject::cutElementByGeometry(const Elem * /*elem*/,
+                                          std::vector<CutEdge> & /*cut_edges*/,
+                                          Real /*time*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+MeshCut3DUserObject::cutElementByGeometry(const Elem * elem,
+                                          std::vector<CutFace> & cut_faces,
+                                          Real /*time*/) const
+// TODO: Time evolving cuts not yet supported in 3D (hence the lack of use of the time variable)
+{
+  bool elem_cut = false;
+
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = _cut_mesh->elements_begin();
+
+  for (unsigned int i = 0; i < elem->n_sides(); ++i)
+  {
+    // This returns the lowest-order type of side.
+    std::unique_ptr<Elem> curr_side = elem->side(i);
+    if (curr_side->dim() != 2)
+      mooseError("In cutElementByGeometry dimension of side must be 2, but it is ",
+                 curr_side->dim());
+    unsigned int n_edges = curr_side->n_sides();
+
+    std::vector<unsigned int> cut_edges;
+    std::vector<Real> cut_pos;
+
+    for (unsigned int j = 0; j < n_edges; j++)
+    {
+      // This returns the lowest-order type of side.
+      std::unique_ptr<Elem> curr_edge = curr_side->side(j);
+      if (curr_edge->type() != EDGE2)
+        mooseError("In cutElementByGeometry face edge must be EDGE2, but type is: ",
+                   libMesh::Utility::enum_to_string(curr_edge->type()),
+                   " base element type is: ",
+                   libMesh::Utility::enum_to_string(elem->type()));
+      Node * node1 = curr_edge->get_node(0);
+      Node * node2 = curr_edge->get_node(1);
+
+      elem_it = elem_begin;
+      for (; elem_it != elem_end; ++elem_it)
+      {
+        std::vector<Point> vertices;
+        Elem * cut_elem = *elem_it;
+
+        for (unsigned int i = 0; i < cut_elem->n_nodes(); i++)
+        {
+          Node * node = cut_elem->get_node(i);
+          // Point * pt = dynamic_cast<Point *>(node);
+          Point & this_point = *node;
+          vertices.push_back(this_point);
+        }
+
+        Point intersection;
+        if (intersectWithEdge(*node1, *node2, vertices, intersection))
+        {
+          cut_edges.push_back(j);
+          cut_pos.push_back(getRelativePosition(*node1, *node2, intersection));
+        }
+      }
+    }
+
+    if (cut_edges.size() == 2)
```

Comment:
I have no idea what you are doing, it'd be nice to see a comment sprinkled in here somewhere.